### PR TITLE
Add prev_hash to soft confirmations

### LINKED ISSUE #845: add prev_hash to soft confirmations
TODOs:

- [ ] Add `prev_hash` to soft confirmations, this should be the hash of the previous soft conrirmation.
  - [ ] This should be included in the hash, signature calc. etc. not just added to the struct.
- [ ] In soft confirmation structs, remove `pre_state_root`, rename `post_state_root` to just `state_root`. 
  - [ ] The `pre_state_root` field in `HookSoftConfirmationInfo` struct can be supplied by the instance of runners itself, rather than being retrieved from the soft confirmation and verified. (Will need a lot of passing around `pre_state_root`)
- [ ] In STF verifier, check that n-1.hash == n.prev_hash while applying soft confirmations.

### PR DESCRIPTION
**Pull request by rakanalh**
# Description

## Linked Issues
- Fixes #845

### CODE REVIEW

**Review by eyusufatik — CHANGES_REQUESTED** [review]
great work. only one problem, we pass state root = 0 to all stf modules in the begin_soft_confirmation_hook now. I mostly addressed this in my comments

File: crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/mod.rs [hunk 1]
```diff
@@ -156,6 +156,15 @@ impl LedgerDB {
...
     }
 
+    /// Gets all soft confirmations by numbers
+    #[instrument(level = "trace", skip(self), err)]
+    pub fn get_soft_batch_by_number(
+        &self,
+        number: &BatchNumber,
+    ) -> Result<Option<StoredSoftBatch>, anyhow::Error> {
+        self.db.get::<SoftBatchByNumber>(number)
+    }
```
[the remark this answers is not in the document]
Comment (rakanalh) [reply crates/sovereign-sdk/full-node/db/sov-db/src/ledger_db/mod.rs:166]: We have it in the RPC handler of the ledger_db... i should change the implementation of the RPC handler to use this

File: crates/sovereign-sdk/module-system/sov-modules-stf-blueprint/src/lib.rs [hunk 2]
```diff
@@ -179,7 +178,6 @@ where
     fn begin_soft_batch(
         &self,
         sequencer_public_key: &[u8],
```
[the remark this answers is not in the document]
Comment (rakanalh) [reply crates/sovereign-sdk/module-system/sov-modules-stf-blueprint/src/lib.rs:181]: Done

File: crates/sovereign-sdk/module-system/sov-modules-api/src/hooks.rs [hunk 3]
```diff
@@ -139,7 +139,7 @@ impl From<SignedSoftConfirmationBatch> for HookSoftConfirmationInfo {
             da_slot_height: signed_soft_confirmation_batch.da_slot_height(),
             da_slot_hash: signed_soft_confirmation_batch.da_slot_hash(),
             da_slot_txs_commitment: signed_soft_confirmation_batch.da_slot_txs_commitment(),
-            pre_state_root: signed_soft_confirmation_batch.pre_state_root(),
+            pre_state_root: vec![0; 32], //TODO signed_soft_confirmation_batch.pre_state_root(),
```
[the remark this answers is not in the document]
Comment (rakanalh) [reply crates/sovereign-sdk/module-system/sov-modules-api/src/hooks.rs]: Done

File: crates/sovereign-sdk/full-node/sov-stf-runner/src/runner.rs
[the remark this answers is not in the document]
Comment (rakanalh) [reply crates/sovereign-sdk/full-node/sov-stf-runner/src/runner.rs:1]: https://github.com/chainwayxyz/citrea/issues/881

File: crates/sovereign-sdk/module-system/sov-modules-api/src/hooks.rs [hunk 4]
```diff
@@ -151,11 +151,11 @@ impl From<SignedSoftConfirmationBatch> for HookSoftConfirmationInfo {
 impl From<HookSoftConfirmationInfo> for SignedSoftConfirmationBatch {
     fn from(val: HookSoftConfirmationInfo) -> Self {
         SignedSoftConfirmationBatch::new(
+            [0u8; 32],
```
[the remark this answers is not in the document]
Comment (rakanalh) [reply crates/sovereign-sdk/module-system/sov-modules-api/src/hooks.rs:157]: I have changed this functionality already into `HookSoftConfirmationInfo::new`

File: crates/sovereign-sdk/rollup-interface/src/state_machine/zk/mod.rs [hunk 5]
```diff
@@ -166,6 +166,8 @@ pub struct StateTransitionData<StateRoot, Witness, Da: DaSpec> {
     pub initial_state_root: StateRoot,
     /// The state root after the state transition
     pub final_state_root: StateRoot,
+    /// The hash before the state transition
+    pub initial_batch_hash: [u8; 32],
```
[the remark this answers is not in the document]
Comment (rakanalh) [reply crates/sovereign-sdk/rollup-interface/src/state_machine/zk/mod.rs:171]: Done

**Review by eyusufatik — APPROVED** [review]

**Review by ercecan — APPROVED** [review]

**Pull request merged**